experimental/sdk/webpubsub-socketio-extension/src/EIO/components/web-pubsub-translator.ts

```diff
@@ -1,11 +1,185 @@
+            debug(`onUserEvent, connectionId = ${connectionId}, handle payload = ${payload}`);
+
+            // @ts-ignore to access private `clients` property
+            var packet = this.linkedEioServer.clients[connectionId].transport.parser.decodePacket(payload); // prettier-ignore
+
+            // @ts-ignore to access private `clients` property
+            this.linkedEioServer.clients[connectionId].onPacket(packet);
+          };
+
+          var payloads = (req.data as string).split(String.fromCharCode(30));
+          for (var i = 0; i < payloads.length; i++) {
+            handlePayload(payloads[i]);
+          }
+          return res.success();
+        } else {
+          return res.fail(401);
+        }
+      },
+
+      onDisconnected: async (req) => {
+        var connectionId = req.context.connectionId;
+        debug(`onDisconnected, connectionId = ${connectionId}`);
+        if (!this.clientConnections.delete(connectionId)) {
+          debug(`onDisconnected, Failed to delete non-existing connectionId = ${connectionId}`);
+        }
+      },
+    });
+  }
+
+  /**
+   * @returns AWPS event handler middleware for EIO Server.
+   */
+  public getEventHandlerEioMiddleware() {
+    /**
+     * AWPS package provides Express middleware for event handlers.
+     * However Express middleware is not compatiable to be directly used by EIO Server.
+     * So a temporary Express App and HttpServer are created as bridges to convert Express middleware to EIO middleware.
+     */
+    const expressMiddleware = this.webPubSubEventHandler.getMiddleware();
+    
+    // Pass Web PubSub Express middlewares into Engine.IO middlewares
+    let bridgeApp = express();
+    bridgeApp.use(expressMiddleware);
+    const bridgeHttpServer = new HttpServer(bridgeApp);
+
+    return bridgeHttpServer.listeners("request")[0];
+  } 
+
+  public getNextSid = () => this.candidateSids.shift();
+
+  /**
+   * Convert an AWPS `connect` request to an Engine.IO `handshake` request.
+   * @param req AWPS `connect` request.
+   * @param context Corrsponding `ClientConnectionContext` for the connecting client. It will be used in `createTransport` to bind each transport to the correct AWPS client connection.
+   */
+  private convertWebPubsubConnectReqToEioHandshakeReq(
+    req: WebPubSubConnectRequest,
+    context: ClientConnectionContext
+  ) {
+    /**
+     * Properties inside are used in EIO src: server.ts `handshake` method.
+     */
+    var dummyEioHandshakeReq: any = {
+      method: "GET", 
+      headers: {},
+      websocket: null,
+      statusCode: null,
+      statusMessage: null,
+      connection: {},
+      url: this.webPubSubOptions.path,
+      _query: { EIO: req.queries.EIO[0], transport: WEBPUBSUB_TRANSPORT_NAME },
+      webPubSubContext: context,
+    };
+    for (var key in req.headers) {
+      let _key = key.toLowerCase();
+      dummyEioHandshakeReq["headers"][_key] = _key == "upgrade" ? req.headers[key][0] : req.headers[key];
```

Comment:
why upgrade is special?

Reply:
`req.headers.Upgrade = ['websocket']` (array), where `req` is the `connect` event request sent by service.
Engine.IO internal method related with WebSocket upgrade processes "req.upgrade" as a string rather than an string array. 

But in current implementation, WebSocket upgrade methods inside Engine.IO won't be triggered. So I will remove this special judge.